server/src/main/java/io/druid/server/SegmentManager.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Ordering;
+import com.google.inject.Inject;
+import com.metamx.emitter.EmittingLogger;
+import io.druid.collections.CountingMap;
+import io.druid.segment.ReferenceCountingSegment;
+import io.druid.segment.Segment;
+import io.druid.segment.loading.SegmentLoader;
+import io.druid.segment.loading.SegmentLoadingException;
+import io.druid.timeline.DataSegment;
+import io.druid.timeline.VersionedIntervalTimeline;
+import io.druid.timeline.partition.PartitionChunk;
+import io.druid.timeline.partition.PartitionHolder;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SegmentManager
+{
+  private static final EmittingLogger log = new EmittingLogger(SegmentManager.class);
+
+  private final Object lock = new Object();
+  private final SegmentLoader segmentLoader;
+  private final Map<String, VersionedIntervalTimeline<String, ReferenceCountingSegment>> dataSources = new HashMap<>();
```

Comment:
Ok for not including this into this PR.

Synchronization that you mentioned still needed could be managed by ConcurrentHashMap:
```java
dataSources.compute(dataSource, dataSourceState -> {
  if (dataSourceState == null) ...
  else {
    dataSourceState.add(...);
  }
});
```
ConcurrentHashMap guarantees that executions of lambdas provided to `compute()`, `computeIfAbsent()`, `merge()` etc. are linearizable. Internally it is implemented via the same intrinsic locks that you use explicitly, but striped over the ConcurrentHashMap entries.